### PR TITLE
Add support for router-wide retry configuration

### LINKED ISSUE #336: make retry budgets configurable
there is a good default retry budget, but we should support overriding it.  backoff strategies should be similarly configurable.

### PR DESCRIPTION
**Pull request by olix0r**
Allows users to configure backoffs and retry budgets.

Fixes #336

### CODE REVIEW

File: linkerd/core/src/test/scala/io/buoyant/linkerd/RouterTest.scala [hunk 1]
```diff
@@ -96,4 +96,21 @@ servers:
...
+         |    backoff:
+         |      kind: jittered
+         |      minMs: 100
+         |      maxMs: 10000
+         |    budget:
+         |      ttlSecs: 30
+         |      minRetriesPerSec: 3
+         |      percentCanRetry: 0.33
+         |""".stripMargin
+    assert(parse(yaml, Stack.Params.empty) != null)
```
Comment (adleong) [inline linkerd/core/src/test/scala/io/buoyant/linkerd/RouterTest.scala]: we should check the value of the `Retries.Budget` stack param in the parsed router.

**Comment by adleong** [comment]
reminder to document this in config.md as part of #364

File: linkerd/docs/config.md [hunk 2]
```diff
@@ -273,6 +273,46 @@ aperture supports the following options (see [here][aperture] for option semanti
...
+      allowed in order to accommodate clients that have just started
+      issuing requests as well as clients that do not issue many
+      requests per window. Must be non-negative and if `0`, then no
+      reserve is given. (Default: 10)
+    * *percentCanRetry* -- Optional. The percentage of calls that can
+      be retried. This is in addition to any retries allowed for via
+      `minRetriesPerSec`.  Must be >= 0 and <= 1000. As an example, if
+      `0.1` is used, then for every 10 non-retry calls , 1 retry will
+      be allowed. If `2.0` is used then every non-retry call, 2
+      retries. (Default: 0.2)
```
Comment (klingerf) [inline linkerd/docs/config.md]: I think this sentence is clearer as: "If 2.0 is used then every non-retry call will be allowed 2 retries."

File: linkerd/docs/config.md [hunk 3]
```diff
@@ -273,6 +273,46 @@ aperture supports the following options (see [here][aperture] for option semanti
...
+    be used (see below).
+    * *kind* -- The name of a backoff algorithm. Either _constant_ or
+      _jittered_.
+
+The _constant_ backoff policy takes a single configuration parameter:
+* _ms_ -- The number of milliseconds to wait before each retry.
+
+The _jittered_ backoff policy uses a
+[decorrelated jitter](http://www.awsarchitectureblog.com/2015/03/backoff.html)
+backoff algorithm and requires takes two configuration parameters:
```
Comment (klingerf) [inline linkerd/docs/config.md]: Typo (no requires): "... and takes two configuration parameters:"

File: linkerd/docs/config.md [hunk 4]
```diff
@@ -273,6 +273,46 @@ aperture supports the following options (see [here][aperture] for option semanti
...
+* _minMs_ -- Optional. The minimum number of milliseconds to wait
+  before each retry.
+* _maxMs_ -- The maximum number of milliseconds to wait before each
+  retry.
```
Comment (klingerf) [inline linkerd/docs/config.md]: As far as I can tell, both of these parameters are actually required, in which case I'd reword the sentence above, and remove "Optional".

Comment (olix0r) [inline linkerd/docs/config.md]: They're not. jackson will happily default them to zero.

File: linkerd/core/src/test/scala/io/buoyant/linkerd/RouterTest.scala [hunk 5]
```diff
@@ -96,4 +102,23 @@ servers:
...
   }
+
+  test("with retries") {
+    val yaml =
+      """|protocol: plain
+         |client:
+         |  retries:
+         |    backoff:
+         |      kind: jittered
+         |      maxMs: 10000
```
Comment (klingerf) [inline linkerd/core/src/test/scala/io/buoyant/linkerd/RouterTest.scala]: Related to my comment below, I'm surprised that this config is valid without minMs specified.  Given that the jittered backoff case class is defined as `JitteredBackoffConfig(minMs: Int, maxMs: Int)`, I thought that implies that both params are required -- what am I missing?

File: linkerd/docs/config.md (hunk 4 above)
Comment (olix0r) [inline linkerd/docs/config.md]: i've added a test, though.

Comment (klingerf) [inline linkerd/docs/config.md]: Ah, in that case I wonder if we'd be better off providing our own defaults.  Something like:

```
case class JitteredBackoffConfig(minMs: Option[Int], maxMs: Option[Int]) extends BackoffConfig {
  def mk = Backoff.decorrelatedJittered(
    start = minMs.map(_.millis).getOrElse(Duration.Zero),
    maximum = maxMs.map(_.millis).getOrElse(100.millis)
  )
}
```

File: linkerd/core/src/test/scala/io/buoyant/linkerd/RouterTest.scala (hunk 5 above)
Comment (olix0r) [inline linkerd/core/src/test/scala/io/buoyant/linkerd/RouterTest.scala]: jackson doesn't enforce any sort of requirements, it will just leave null-values for missing fields. for Ints, this is 0 (which is good for minMs).  for options, it's None.

**Comment by olix0r** [comment]
@klingerf good call.  writing better tests for this showed that both min and max have to be greater than 0, so they're both required.

**Comment by klingerf** [comment]
⭐ 🍧 Great, thanks for making those updates.

File: linkerd/core/src/test/scala/io/buoyant/linkerd/RouterTest.scala (hunk 5 above)
Comment (adleong) [inline linkerd/core/src/test/scala/io/buoyant/linkerd/RouterTest.scala]: This test is a bit confusing.  It creates a RetriesConfig that will fail if `.mk` is called on it.  Since you test that error case elsewhere, I'd just give a fully valid config here.

Comment (olix0r) [inline linkerd/core/src/test/scala/io/buoyant/linkerd/RouterTest.scala]: ah yeah, that's an oversight. good catch.

**Comment by gtcampbell** [comment]
:ship: looks nice, and I agree that the tests are very helpful.
